### PR TITLE
feat(kustomize): add patches field (#5114)

### LINKED ISSUE #5114: Add support to argoproj/Application for kustomize.patchesJson6902 (and patchesStrategicMerge)
# Summary

Application can specify some kustomize parameters (namePrefix, nameSuffix, images) but not all of them. I would like support adding for patchesJson6902 and whilst in there patchesStrategicMerge which could be used with inlinePatches (https://github.com/kubernetes-sigs/kustomize/blob/master/examples/inlinePatch.md).

I prefer having strategicMerge and json6902 separate rather than a generic 'patches' option but am open to debate.

# Motivation

I would like to deploy external-dns (https://github.com/kubernetes-sigs/external-dns) to all of my clusters. In the interests of DRY I would like a single kustomization bundle which configures all the clusters that argo manages essentially the same. The argument --text-owner-id however should be different for each one. I was hoping to json6902 "op: add" that using a separate 'application' all pointing at the same git repository, with just an inline patch to add that single parameter.

# Proposal

Add the appropriate patchesJson6902 and patchesStrategicMerge to the Application CRD (types.go) and in AddAppFlags and then honor those when we call kustomize.Build.
Would also want GUI support in application-parameters.tsx

If this is something people want then I'm happy to tackle it.

### PR DESCRIPTION
**Pull request by crenshaw-dev**
This still needs tests.

Fixes #5114

### CODE REVIEW

**Comment by acelinkio** [comment]
Hey @crenshaw-dev.  I am not quite prepared to build this project.  I can open an MR into your branch for adding some docs and prepare some tests. 

Is there a list of things you'd like to have addressed?

**Comment by acelinkio** [comment]
Created PR to your branch for adding doc + example.  https://github.com/crenshaw-dev/argo-cd/pull/20

**Comment by acelinkio** [comment]
Looked into adding tests and quickly realized cli support is needed.  I found where the code should be added, but not experienced enough as a Go developer to take on that challenge.  Kustomize data structures are complicated and could not find a useful reference in this project.

Test: https://github.com/argoproj/argo-cd/blob/master/test/e2e/kustomize_test.go
CLI:  https://github.com/argoproj/argo-cd/blob/master/cmd/util/app.go#L223

**Comment by jtcressy** [comment]
> Thanks, @acelinkio, merged the docs!
> 
> CLI support might be weird/tricky since the patches field is an array, and the contents might be large (awkward to provide as CLI params). Maybe accept the whole `patches` field as a YAML string or file?

What if some behavior of the `kustomize` CLI interface could be mimicked? Current usage of kustomize CLI is like so: `kustomize edit add patch --path {filepath} --group {target group name} --version {target version}`

Also, `kustomize edit add patch` supports a `--patch` argument that expects a literal string of the patch to apply so you may be able to fully implement the CLI interface for both usage methods assuming no other complications.

**Comment by acelinkio** [comment]
I can see that but not sure what the long term direction of kustomize plugin will be.

If everything is able to be done inline, there is no need to clone a repository.  Any resources could be specified via resource.

I just finished creating an quick plugin for just that.  Because some projects are only publishing to GitHub releases.

```
echo "$ARGOCD_ENV_KUSTOMIZATION_YAML" > kustomization.yaml;
kustomize build
```

**Comment by sourceful-karlson** [comment]
is there any thing we can do to move this PR forwards? This is a blocker for a feature that we have.

**Comment by sourceful-karlson** [comment]
If not, what is a workaround on this?

**Comment by crenshaw-dev** [comment]
@sourceful-karlson at this point we need tests and, ideally, CLI/UI support. If you're up to open a PR against this one, that would help.

We're looking at [2.9 (Nov 6)](https://argo-cd.readthedocs.io/en/latest/developer-guide/release-process-and-cadence/) at best, so just keep that in mind.

**Comment by sourceful-karlson** [comment]
Screenshot 2023-08-11 at 15 55 32
FYI this is for doing kustomise patch in the applicationset for pull request

**Comment by sourceful-karlson** [comment]
on the CLI support, would it make sense to use JSON? the same way how kubectl patch works?

**Comment by crenshaw-dev** [comment]
@sourceful-karlson yep! I think sticking as close as possible to Kustomize behavior makes sense.

**Comment by sourceful-karlson** [comment]
just want to confirm what is the minimum list of things to work for this to get merged

**Comment by crenshaw-dev** [comment]
In my opinion, we can push forward by just adding tests to what's here already. UI/CLI can wait for someone who has time to add those.

**Comment by sathieu** [comment]
@crenshaw-dev Thanks for this PR!

My usecase is using patches from an Helm application. Is this covered?

**Comment by sourceful-karlson** [comment]
we are getting closed to the release candidate for this, does this mean this PR will miss the boat?

**Comment by sourceful-karlson** [comment]
I will try to tackle this test thing this thursday

**Comment by crenshaw-dev** [comment]
@sourceful-karlson this will only cover your use case if the Helm chart is wrapped in a Kustomize overlay. This uses the kustomize CLI's `patches` functionality - it does not re-implement the functionality for all Argo CD apps regardless of source type. I think that would be a good future enhancement.

I've asked to push 2.9 RC1 to next Monday to provide more time to merge features. If you're able to add tests, and I'm able to get an approving review before then, it'll be in 2.9. :-)

**Comment by sourceful-karlson** [comment]
Beginner Open Source Developer

https://github.com/crenshaw-dev/argo-cd/pull/21/files

Early draft for anything wrong that could be catched early

**Comment by crenshaw-dev** [comment]
Thanks, @sourceful-karlson, looking great! Just lmk when it's ready for review.

**Comment by sourceful-karlson** [comment]
ready-ish, have a look?

**Comment by sourceful-karlson** [comment]
is there anything I can do to move to forwards?

**Comment by crenshaw-dev** [comment]
We'll need a review from another maintainer.

Do you have time to look into this e2e test failure? Looks related. https://github.com/argoproj/argo-cd/actions/runs/6262996312/job/17006646003?pr=14648

**Comment by sourceful-karlson** [comment]
Doesn't look related, the patch is http patch call, not our patch I think

https://github.com/argoproj/argo-cd/commit/e08840f12b627d8a1ebf02a4ddf4258601192f3f#diff-78f42ba40d0f10b08c73b7e6bb8376f398e249c963cf549e89591d6b6826b9a4

**Comment by sourceful-karlson** [comment]
@alexec @terrytangyuan got time for a review?

**Review by ishitasequeira — COMMENTED** [review]
Overall PR looks good!! Just some minor comments

File: util/kustomize/kustomize.go
Comment (crenshaw-dev) [reply util/kustomize/kustomize.go]: Done!

File: util/kustomize/kustomize.go
Comment (crenshaw-dev) [reply util/kustomize/kustomize.go:239]: The appropriate locking is done a few levels up by the `runRepoOperation` method in repository.go.

**Review by ishitasequeira — APPROVED** [review]
LGTM!!

**Comment by sourceful-karlson** [comment]
@crenshaw-dev is this in the rc?

**Comment by sourceful-karlson** [comment]
@leoluz can I check that this feature will be in the release?

**Comment by blakepettersson** [comment]
@sourceful-karlson https://github.com/argoproj/argo-cd/commits/v2.9.0-rc1

TLDR - it's there

**Comment by crenshaw-dev** [comment]
@sourceful-karlson or rather it will be there. The RC build failed yesterday evening. Looking into it this morning. :-)

**Comment by sourceful-karlson** [comment]
@crenshaw-dev noob quesiton, is rc-2 ready to be used and tested at my own risk?

**Comment by crenshaw-dev** [comment]
@sourceful-karlson yep! From what I've seen, the RCs are pretty stable. Maybe just a few UI bugs.

**Comment by sourceful-karlson** [comment]
I am testing this feature, it doesnt seems like the patch is working

**Comment by sourceful-karlson** [comment]
sorry, confirmed that it s working

**Comment by sourceful-karlson** [comment]
was patching with common label but realised that it is not right, as common label is not there during patching

**Comment by sourceful-karlson** [comment]
@crenshaw-dev just wanna say, this is a killer feature, and we basically able to have deployment preview on our pull request due to this feature.

applicationset should be advertised more

**Pull request merged**
